zilliqa/src/exec.rs

```diff
@@ -1296,176 +1318,216 @@ fn scilla_create(
     ))
 }
 
-fn scilla_transfer_to_eoa(
-    mut state: PendingState,
+fn scilla_call(
+    state: PendingState,
+    scilla: MutexGuard<'_, Scilla>,
     from_addr: Address,
     txn: TxZilliqa,
     mut inspector: impl ScillaInspector,
 ) -> Result<(ScillaResult, PendingState)> {
-    let gas = txn.gas_limit;
-
-    let Some(gas) = gas.checked_sub(SCILLA_TRANSFER) else {
-        warn!("not enough gas to make transfer");
-        return Ok((
-            ScillaResult {
-                success: false,
-                contract_address: None,
-                logs: vec![],
-                gas_used: (txn.gas_limit - gas).into(),
-                accepted: Some(false),
-                errors: [(0, vec![ScillaError::OutOfGas])].into_iter().collect(),
-                exceptions: vec![],
-            },
-            state,
-        ));
+    let mut gas = txn.gas_limit;
+
+    let message = if !txn.data.is_empty() {
+        let mut m: Value = serde_json::from_str(&txn.data)?;
+        m["_amount"] = txn.amount.to_string().into();
+        m["_sender"] = format!("{from_addr:#x}").into();
+        m["_origin"] = format!("{from_addr:#x}").into();
+        Some(m)
+    } else {
+        None
     };
 
-    if let Some(result) = state.deduct_from_account(from_addr, txn.amount)? {
-        return Ok((result, state));
-    }
+    let mut call_stack = vec![(0, from_addr, txn.to_addr, txn.amount, message)];
+    let mut logs = vec![];
+    let mut transitions = vec![];
+    let mut root_contract_accepted = false;
 
-    let to = state.load_account(txn.to_addr)?;
-    to.account.balance += txn.amount.get();
+    let mut state = Some(state);
 
-    inspector.transfer(from_addr, txn.to_addr, txn.amount.get());
+    while let Some((depth, sender, to_addr, amount, message)) = call_stack.pop() {
+        let mut current_state = state.take().expect("missing state");
 
-    Ok((
-        ScillaResult {
-            success: true,
-            contract_address: None,
-            logs: vec![],
-            gas_used: (txn.gas_limit - gas).into(),
-            accepted: None,
-            errors: BTreeMap::new(),
-            exceptions: vec![],
-        },
-        state,
-    ))
-}
+        let code_and_data = match &current_state.load_account(to_addr)?.account.code {
+            // Both EOAs and EVM contracts are represented by [Code::Evm]. For now we handle them all as EOAs.
+            Code::Evm(_) => None,
```

Comment:
I guess this will change with interop. But I am curious - Do you know the current behaviour of ZQ1 if a Scilla contract sends a message to an EVM contract with value? My assumption was that it would behave like an EOA, since we have no Scilla->EVM interop.

Reply:
Yep. Let me check - I think you get an error, but ..

Reply:
Yep you're right:

* Sending funds to an EVM contract: https://otterscan.testnet.zilliqa.com/tx/0x1c4281ff205b9b6c0f3055144b605d83adea79d0555b7919b6005afe4b8847f5
* Sending funds to a Scilla contract (without a matching transition): https://otterscan.testnet.zilliqa.com/tx/0xd1f0925f8a2bb49ac9bb4ed70c61bfc7067cd307dbe52a197100bdc58d9a056e
* Sending funds to an EOA: https://otterscan.testnet.zilliqa.com/tx/0x322b59118c108074f45948dbacae9bfed2b794ce986426419a76dfe06bfa5798

I will update this PR to reject transfers to EVM accounts that aren't empty (i.e. aren't EOAs).

Reply:
Has to happen once a decade or so, just by chance :-) - thank you for checking & good luck! Yell when you want a re-review ..

Reply:
Ready for re-review now :)